APPS-32904: Allow reusing a connection from another connector via session and master token

sfc-gh-klin opened the pull request:
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Addresses APPS-32904

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code (haven't checked out tests, in-progress)
   - [ ] I am adding new logging messages (shouldn't be needed?) 
   - [ ] I am adding a new telemetry message (shouldn't be needed?) 
   - [x] I am modifying authorization mechanisms (authentication) 
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

The Snowflake VS Code extension uses the node connector (snowflake-sdk) to create a connection. Users who use Snowpark in conjunction with the VS Code extension would benefit if the extension lets them use the node connector's connection that they've already authenticated with by logging in through the extension UI. Then when they execute Snowpark code, they wouldn't have to create another connection through the Python connector. 

To get started with this flow, the Python connector needs to be able to accept a session and master token from another connection. 

I'm interested in starting some discussion around this via this PR. 

4. Considerations

`update_tokens` takes some other parameters such as id_token and mfa_token. The Node connector's connection, when serialized, does not have that information. I tested with a trial Snowflake account with MFA enabled and I successfully logged in and reused the node connector connection in Python code. I'm not 100% clear on whether omitting these (id_token, mfa_token) could affect anything else. In connection.py's open_connection, by using the session and master token in update_tokens, I may be bypassing some validation that occurs in all the other authentication paths. Interested in learning more from the regular contributors to this repository.

sfc-gh-klin commented:
I have read the CLA Document and I hereby sign the CLA

sfc-gh-klin commented:
recheck

sfc-gh-mkeller commented on src/snowflake/connector/connection.py:
```diff
@@ -913,100 +925,108 @@ def __open_connection(self):
-                self._session_parameters,
+
+        if self._session_token and self._master_token:
+            auth._rest.update_tokens(
```
I think we should test whether the "authentication" was successful somehow

sfc-gh-klin replied:
Okay, if I try to make a heartbeat call there and then it fails, user will get this error in their terminal for example
```
Traceback (most recent call last):
  File "/Users/klin/git/sp-test/src/sp_test.py", line 3, in <module>
    conn = snowflake.connector.connect(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/__init__.py", line 54, in Connect
    return SnowflakeConnection(**kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/connection.py", line 425, in __init__
    self.connect(**kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/connection.py", line 715, in connect
    self.__open_connection()
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/connection.py", line 935, in __open_connection
    ret = auth._rest._heartbeat()
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 609, in _heartbeat
    ret = self._post_request(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 730, in _post_request
    ret = self.fetch(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 839, in fetch
    ret = self._request_exec_wrapper(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 975, in _request_exec_wrapper
    raise e
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 880, in _request_exec_wrapper
    return_object = self._request_exec(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 1189, in _request_exec
    raise err
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 1131, in _request_exec
    raise_failed_request_error(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/network.py", line 238, in raise_failed_request_error
    Error.errorhandler_wrapper(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/errors.py", line 290, in errorhandler_wrapper
    handed_over = Error.hand_to_other_handler(
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/errors.py", line 348, in hand_to_other_handler
    connection.errorhandler(connection, cursor, error_class, error_value)
  File "/opt/homebrew/lib/python3.9/site-packages/snowflake/connector/errors.py", line 221, in default_errorhandler
    raise error_class(
snowflake.connector.errors.InterfaceError: 250003 (08001): 401 Unauthorized: post https://ig86763.preprod12.us-west-2.aws.snowflakecomputing.com:443/session/heartbeat?requestId=4a63f07e-4928-4f98-b02d-58a4e7151f73&request_guid=148cac21-a4da-4341-bb24-862f8a6211c7
```

Does that seem like a reasonable experience to give a dev for when the tokens are invalid? Should that error get caught and replaced with clearer wording?

sfc-gh-mkeller commented on src/snowflake/connector/connection.py:
```diff
@@ -1115,7 +1135,12 @@ def __config(self, **kwargs):
 
-        if not self.user and self._authenticator != OAUTH_AUTHENTICATOR:
+        if (
+            not self._master_token
```
I feel like instead of modifying this and the next if statements it'd be better to rearrange the code so that error checking (btw you missed the account one) could simply be indented with the other old code.

sfc-gh-klin replied:
Okay, rearranged! Let me know if that's what you were suggesting

sfc-gh-mkeller reviewed (changes requested):
We should also discuss testing plan for this! I think we could do a integration test where we extract these from a normally created connection.

sfc-gh-klin commented:
I added an integration test called test_with_tokens and ran it with `python -m pytest test/integ/test_connection.py::test_with_tokens` after adding a parameters.py file filled out. 

However, I'm puzzled as to whether or not my test genuinely tests what I am trying to do, because I can put nonsense in for the session_token and master_token, and as long as I have the account/user/password/host filled out it'll pass. I think this is why you suggested to check that the "authentication" succeeded since otherwise it's just supplying the session/master token and storing those to be used when user actually tries to execute something. 

The other thing I'm confused by is that I would expect to be able to provide only the account, master_token, and session_token, but if in parameters.py I don't provide user/password then the integration test also does not succeed. 

```
ERROR test/integ/test_connection.py::test_with_tokens - snowflake.connector.errors.DatabaseError: 250001 (08001): None: Failed to connect to DB: ig86763.preprod12.us-west-2.aws.snowfl...
```
I see that this is because we do some setup in conftest.py around init_test_schema for example, and that connection uses user/password. Is it reasonable for me to also pass along session_token and master_token to give flexibility with parameters.py usage?

sfc-gh-klin commented:
Huh, kind of interesting. Trying to use session_token and master_token by editing init_test_schema with 
```
    ret = db_parameters
    with snowflake.connector.connect(
        user=ret["user"],
        password=ret["password"],
        host=ret["host"],
        port=ret["port"],
        database=ret["database"],
        account=ret["account"],
        protocol=ret["protocol"],
        session_token=ret["session_token"],
        master_token=ret["master_token"]
```
doesn't seem to work because it immediately makes the token expired. 

```
ERROR test/integ/test_connection.py::test_with_tokens - snowflake.connector.errors.ProgrammingError: 390111: 390111: Session no longer exists.  New login required to access the service.
```

For some reason it's okay when I reuse the token myself in a Python file I execute (which is how I manually tested my code changes) but trying to do it through conftest.py expires the token and my other uses of the connection also die.

sfc-gh-klin commented:
Removed my previous comments to avoid confusion, I've rewritten the unit test to basically make a first connection then take the tokens off of it and create a second connection. Currently I'm having a local failure on `python -m pytest test/integ/test_connection.py::test_ocsp_mode_insecure` and I'll look into. Not sure I understand that test yet.

sfc-gh-aling commented on src/snowflake/connector/connection.py:
```diff
@@ -1115,34 +1138,35 @@ def __config(self, **kwargs):
-                ProgrammingError,
-                {"msg": "User is empty", "errno": ER_NO_USER},
-            )
+        if not self._master_token and not self._session_token:
```
why do we need check to master and session token here?

sfc-gh-klin replied:
In my understanding: Master and session token being provided should override the other authentication methods. 

So we check that if it's not provided, then we do the normal authentication workflow like it was before any code change.

sfc-gh-aling replied:
in this case, should the condition be `if not (self._master_token and self._session_token)` in which case we only skip the normal authentication workflow when both tokens are detected?

otherwise if only one token is provided, we shall use the normal authentication workflow

sfc-gh-klin replied:
Thanks for catching that, that makes sense to me.

sfc-gh-aling commented on src/snowflake/connector/connection.py:
```diff
@@ -913,100 +925,111 @@ def __open_connection(self):
-                    protocol=self._protocol,
-                    host=self.host,
-                    port=self.port,
+            auth._rest._heartbeat()
```
I'm curious what kind of error will be raised if invalid/expired token is provided?

sfc-gh-klin replied:
This is what it shows:
```snowflake.connector.errors.InterfaceError: 250003 (08001): 401 Unauthorized: post https://ig86763.preprod12.us-west-2.aws.snowflakecomputing.com:443/session/heartbeat?requestId=4a63f07e-4928-4f98-b02d-58a4e7151f73&request_guid=148cac21-a4da-4341-bb24-862f8a6211c7```

I think I can try to write a test for that.

sfc-gh-aling replied:
yeah let's add a test for it -- error behavior is also part of the public apis

sfc-gh-klin replied:
I added a test for it and initially I edited errors.py to add another error type (the Unauthorized one) but I noticed it caused issues with the old driver test due to imports, so I instead went with ProgrammingError since technically speaking the tokens aren't valid anymore. What's your take on this?

sfc-gh-aling replied:
I don't think we need to introduce another error type, whatever is returned nowadays shall be good

sfc-gh-klin replied:
Okay, roger that.

sfc-gh-aling commented on src/snowflake/connector/connection.py:
```diff
@@ -256,6 +256,18 @@ def _get_private_bytes_from_file(
+        None,
+        (type(None), str),
+    ),  # master token from another connection, to be provided together with session token
+    "master_validity_in_seconds": (
```
does the session token has a similar validity in seconds concept?

sfc-gh-klin replied:
I don't think it does--or if it does, it's not something the connector code ever sees. This value I get from the node connector when I serialize the object to get the session, master, and master validity. It has no other info on it. 

Example from the python connector code, it references master validity but nothing about session. 
https://sourcegraph.c2.us-central1.gcp-dev.app.snowflake.com/github.com/snowflakedb/snowflake-connector-python/-/blob/src/snowflake/connector/network.py?L408:16

sfc-gh-aling commented:
@sfc-gh-klin  can you also update the changelog file: https://github.com/snowflakedb/snowflake-connector-python/blob/main/DESCRIPTION.md -- this is considered a new feature to the connector.

sfc-gh-klin commented:
I've updated the changelog, and the failing Jenkins tests are related to a different commit and not my changes--some discussion in this Slack channel https://snowflake.slack.com/archives/C7Y5W9XR6/p1704846477609349?thread_ts=1704840824.860299&cid=C7Y5W9XR6

sfc-gh-aling commented on test/integ/test_connection.py:
```diff
@@ -131,6 +131,63 @@ def test_with_config(db_parameters):
+    session_token = initial_cnx._rest._token
+    initial_cnx.close()
+
+    with pytest.raises(Error):
```
is this throwing the `InterfaceError` based upon our previous discussion?

> snowflake.connector.errors.InterfaceError: 250003 (08001): 401 Unauthorized: post https://ig86763.preprod12.us-west-2.aws.snowflakecomputing.com:443/session/heartbeat?requestId=4a63f07e-4928-4f98-b02d-58a4e7151f73&request_guid=148cac21-a4da-4341-bb24-862f8a6211c7

sfc-gh-klin replied:
It is now a ProgrammingError--I have made the test be more specific and changed it to ProgrammingError.

sfc-gh-aling reviewed (approved):
LGTM

sfc-gh-mkeller commented on src/snowflake/connector/network.py:
```diff
@@ -595,7 +595,7 @@ def _token_request(self, request_type):
             )
 
-    def _heartbeat(self) -> None:
+    def _heartbeat(self) -> Any | dict[Any, Any] | dict[str, Any] | None:
```
`dict[Any, Any]` includes `dict[str, Any]`. No need to have both!
```suggestion
    def _heartbeat(self) -> Any | dict[Any, Any] | None:
```

sfc-gh-mkeller reviewed (approved):
One tiny type-hint thing, please accept that suggestion. Otherwise this looks great!
Thanks for adding tests too! I made them a little safer, hope you don't min!

:ship: :ship:

sfc-gh-klin commented:
Thank you for making the tests safer, this is useful learning for me. 
Accepted your type hint change, will merge when checks pass.
